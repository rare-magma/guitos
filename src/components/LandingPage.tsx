import { Button, Container, Form, Row, Spinner, Stack } from "react-bootstrap";
import { Budget } from "./Budget";
import { RefObject } from "react";

interface LandingPageProps {
  loading: boolean;
  budget: Budget | null;
  budgetList: Budget[];
  inputRef: RefObject<HTMLInputElement>;
  onNew: () => void;
  onImport: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

function LandingPage({
  loading,
  budget,
  budgetList,
  inputRef,
  onNew,
  onImport,
}: LandingPageProps) {
  const handleNew = () => {
    onNew();
  };

  const handleImport = (e: React.ChangeEvent<HTMLInputElement>) => {
    onImport(e);
  };

  return (
    <>
      {loading && (
        <Container
          fluid
          className="position-absolute top-50 start-50 translate-middle"
        >
          <Row className="justify-content-center">
            <Spinner animation="border" role="status" />
          </Row>
        </Container>
      )}

      {!loading && !budget && budgetList.length < 1 && (
        <Container className="position-absolute top-50 start-50 translate-middle">
          <Row className="justify-content-center align-content-center">
            <Stack gap={3}>
              <Button
                className="w-25 align-self-center"
                aria-label="new budget"
                variant="outline-success"
                onClick={handleNew}
              >
                new
              </Button>
              <Form.Group className="w-25 align-self-center" controlId="import">
                <Button
                  className="w-100"
                  aria-label="import budget"
                  variant="outline-primary"
                  onClick={() => {
                    inputRef.current?.click();
                  }}
                >
                  import
                </Button>
                <Form.Control
                  type="file"
                  multiple
                  ref={inputRef}
                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                    handleImport(e);
                  }}
                  style={{ display: "none" }}
                />
              </Form.Group>
            </Stack>
          </Row>
        </Container>
      )}
    </>
  );
}

export default LandingPage;
