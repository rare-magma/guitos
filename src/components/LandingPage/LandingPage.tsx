import { RefObject } from "react";
import { Button, Container, Form, Row, Stack } from "react-bootstrap";
import { useBudget } from "../../context/BudgetContext";
import Loading from "../Loading/Loading";

interface LandingPageProps {
  loading: boolean;
  inputRef: RefObject<HTMLInputElement>;
  onNew: () => void;
  onImport: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

function LandingPage({ loading, inputRef, onNew, onImport }: LandingPageProps) {
  const { budget, budgetList } = useBudget();
  const showLandingPage =
    !loading && !budget && budgetList && budgetList.length < 1;

  function handleNew() {
    onNew();
  }

  function handleImport(e: React.ChangeEvent<HTMLInputElement>) {
    onImport(e);
  }

  return (
    <>
      {loading && <Loading />}

      {showLandingPage && (
        <Container className="position-absolute top-50 start-50 translate-middle">
          <Row className="justify-content-center align-content-center">
            <Stack gap={3}>
              <Button
                className="w-25 align-self-center"
                aria-label="new budget"
                variant="outline-success"
                onClick={handleNew}
              >
                new
              </Button>
              <Form.Group className="w-25 align-self-center" controlId="import">
                <Button
                  className="w-100"
                  aria-label="import budget"
                  variant="outline-primary"
                  onClick={() => {
                    inputRef.current?.click();
                  }}
                >
                  import
                </Button>
                <Form.Control
                  type="file"
                  data-testid="import-form-control-landing-page"
                  multiple
                  ref={inputRef}
                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                    handleImport(e);
                  }}
                  style={{ display: "none" }}
                />
              </Form.Group>
              <Button
                className="w-25 align-self-center"
                aria-label="open instructions in new tab"
                variant="outline-info"
                href="https://github.com/rare-magma/guitos#getting-started"
                target="_blank"
              >
                help
              </Button>
            </Stack>
          </Row>
        </Container>
      )}
    </>
  );
}

export default LandingPage;
