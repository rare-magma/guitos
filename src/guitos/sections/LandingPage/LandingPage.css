.version {
  color: var(--comment);
}

.balanced {
  text-wrap: pretty;
}
