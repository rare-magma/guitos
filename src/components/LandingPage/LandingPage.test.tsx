import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import {
  budgetContextSpy,
  testBudget,
  testEmptyBudgetContext,
} from "../../setupTests";
import LandingPage from "./LandingPage";

describe("LandingPage", () => {
  const inputRefMock: { current: HTMLInputElement | null } = { current: null };
  const onNew = vi.fn();
  const onImport = vi.fn();
  const comp = (
    <LandingPage
      loading={false}
      inputRef={inputRefMock}
      onNew={onNew}
      onImport={onImport}
    />
  );

  beforeEach(() => {
    budgetContextSpy.mockReturnValue(testEmptyBudgetContext);
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    expect(screen.getByLabelText("new budget")).toBeInTheDocument();
    expect(screen.getByLabelText("import budget")).toBeInTheDocument();
    expect(
      screen.getByLabelText("open instructions in new tab"),
    ).toBeInTheDocument();
  });

  it("triggers onNew", async () => {
    const newButton = screen.getAllByRole("button", { name: "new budget" })[0];
    await userEvent.click(newButton);
    expect(onNew).toBeCalledTimes(1);
  });

  it("triggers onImport", async () => {
    await userEvent.upload(
      screen.getByTestId("import-form-control-landing-page"),
      testBudget as unknown as File,
    );
    expect(onImport).toBeCalledTimes(1);
  });

  it("opens instructions in new tab", async () => {
    const instructionsButton = screen.getByLabelText(
      "open instructions in new tab",
    );
    await userEvent.click(instructionsButton);
    expect(instructionsButton).toHaveAttribute(
      "href",
      "https://github.com/rare-magma/guitos#getting-started",
    );
  });

  it("renders loading spinner", () => {
    render(
      <LandingPage
        loading={true}
        inputRef={inputRefMock}
        onNew={onNew}
        onImport={onImport}
      />,
    );
    expect(screen.getByRole("status")).toBeInTheDocument();
  });
});
