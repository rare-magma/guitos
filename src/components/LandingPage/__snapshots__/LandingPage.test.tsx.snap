// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LandingPage > matches snapshot 1`] = `
<LandingPage
  inputRef={
    {
      "current": <input
        class="form-control"
        data-testid="import-form-control-landing-page"
        id="import"
        multiple=""
        style="display: none;"
        type="file"
      />,
    }
  }
  loading={false}
  onImport={[MockFunction spy]}
  onNew={[MockFunction spy]}
/>
`;
